### PR TITLE
Add label editor to create custom labels

### PR DESCRIPTION
**Pull request by serk12**
- Label editor
- fix transparent labels

Signed-off-by: Marc Prat Masó <marc.prat.maso@estudiantat.upc.edu>

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

File: avogadro/qtplugins/label/labeleditor.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    e->accept();
+    if (m_selectedAtom.isValid()) {
+      save();
+    }
+
+    Identifier clickedObject = m_renderer->hit(e->pos().x(), e->pos().y());
+    m_selected = (clickedObject.type == Rendering::AtomType);
+    if (m_selected) {
+      m_selectedAtom = m_molecule->atom(clickedObject.index);
+      m_text = QString::fromUtf8(m_selectedAtom.label().c_str());
```
Comment (ghutchis) [inline avogadro/qtplugins/label/labeleditor.cpp:88]: QString::fromStdString() should also work, I think...

**Comment by ghutchis** [comment]
This Ubuntu failure just seems to be a timeout. (The test set seems to be failing on bond tests #735.)

**Comment by ghutchis** [comment]
I'll test the UI on this. 

In the previous version, if you were showing the label display, it would look like you'd put a new label on top of the label  display, which was weird. I think a user would expect clicking would offer a default (e.g., the "H3" or "C1" from the label display if nothing is set) so that text could be changed.

The label engine probably needs to offer a popup menu / QComboBox with a few alternatives including "Custom Labels Only".

File: avogadro/qtplugins/label/labeleditor.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      save();
+    }
+    emit drawablesChanged();
+  }
+  return nullptr;
+}
+
+void LabelEditor::save()
+{
+  m_molecule->beginMergeMode(tr("write Label"));
```
[the remark this answers is not in the document]
Comment (serk12) [reply avogadro/qtplugins/label/labeleditor.cpp:66]: +1 to "Create Label"

File: avogadro/qtplugins/label/labeleditor.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+namespace {
+TextLabel3D* createLabel(const std::string& text, const Vector3f& pos,
+                         float radius)
+{
+  Rendering::TextProperties tprop;
+  tprop.setAlign(Rendering::TextProperties::HCenter,
+                 Rendering::TextProperties::VCenter);
+  tprop.setFontFamily(Rendering::TextProperties::SansSerif);
+
+  tprop.setColorRgb(255, 255, 255);
```
[the remark this answers is not in the document]
Comment (serk12) [reply avogadro/qtplugins/label/labeleditor.cpp:105]: This is only in the label edit mode not the render display. But maybe white could be annoying if the background is white?

**Comment by serk12** [comment]
Something like this? I will change the checbox with the QComboBox and add the same options to the residues. 
image

**Comment by ghutchis** [comment]
Yes, I think the checkboxes are types of atom labels. So you could probably have something like:

Atom Labels: (as a combo box)
- None
- Atom Index
- Element Symbol
- Symbol & Number in Group
- Custom

(and "Custom" => `atom.label()`)

(Sorry, the font is weird in Avo1 on my Mac now.)
Screen Shot 2021-08-20 at 2 50 29 PM

**Comment by ghutchis** [comment]
We never had residue labels, but probably:

Residue Labels:
- None
- Residue Name
- Residue ID
- Residue Name + ID

I've definitely seen cases where people annotate 2-3 particular residues with labels.

File: avogadro/qtplugins/label/labeleditor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghutchis) [reply avogadro/qtplugins/label/labeleditor.cpp:105]: The white can be annoying for light-colored atoms (e.g., hydrogen)

**Comment by ghutchis** [comment]
Let me know when you think this is ready to merge. The code looks good - just trying to get the UI / UX right. 😄

File: avogadro/qtplugins/label/label.cpp [hunk 4]
```diff
@@ -131,19 +143,74 @@ struct LayerLabel : Core::LayerData
...
+              residue->setCurrentText(text);
+            }
+          }
+        }
+      }
+      QObject::connect(residue, SIGNAL(currentIndexChanged(int)), slot,
+                       SLOT(residueLabelType(int)));
+      index = residue->findData(int(residueOptions));
+      residue->model()->sort(0, Qt::AscendingOrder);
+      form->addRow(QObject::tr("residue label:"), residue);
```
Comment (ghutchis) [inline avogadro/qtplugins/label/label.cpp:211]: tr("Residue Label:")

File: avogadro/qtplugins/label/label.cpp [hunk 5]
```diff
@@ -131,19 +143,74 @@ struct LayerLabel : Core::LayerData
...
+          if (val == atomOptions) {
+            atom->setCurrentText(text);
+          }
+        }
+      }
+      QObject::connect(atom, SIGNAL(currentIndexChanged(int)), slot,
+                       SLOT(atomLabelType(int)));
+      int index = atom->findData(int(atomOptions));
+      atom->model()->sort(0, Qt::AscendingOrder);
+      form->addRow(QObject::tr("Atom label:"), atom);
```
Comment (ghutchis) [inline avogadro/qtplugins/label/label.cpp:181]: tr("Atom Label:")

**Comment by ghutchis** [comment]
I think there are two *minor* issues with the latest version.
- I don't think we need to offer so many options - many users are students. For now, I think I'd stick to the options I mentioned above (https://github.com/OpenChemistry/avogadrolibs/pull/740#issuecomment-902890438). It's obviously easy to add more, but I suspect these are the most common.
- For translation / i18n doing programmatic assembly is bad. European languages work like tr("Element") + tr("Index") but it's much better to translate each assembled string as a literal. (I learned this lesson the hard way .. one benefit of 25+ translations of Avo1.)

File: avogadro/qtplugins/label/label.cpp [hunk 6]
```diff
@@ -131,19 +143,74 @@ struct LayerLabel : Core::LayerData
...
-      check->setChecked(residueLabel);
-      QObject::connect(check, &QCheckBox::clicked, slot, &Label::residueLabel);
-      v->addWidget(check);
+      QComboBox* atom = new QComboBox;
+      atom->setObjectName("atom");
+      for (char i = 0x00; i < std::pow(2, 4); ++i) {
+        if (i == 0) {
+          atom->addItem(QObject::tr("None"), QVariant(LabelOptions::None));
+        } else {
+          char val = 0x00;
```
Comment (ghutchis) [inline avogadro/qtplugins/label/label.cpp:152]: Just use a string list. I know this works in English, but it won't work for translation in many languages.

QStringList options;
options <<  tr("None") << tr("Index") << tr("Element Symbol") << tr("Symbol + Index"); // etc

File: avogadro/qtplugins/label/label.cpp [hunk 7]
```diff
@@ -228,19 +303,32 @@ void Label::processAtom(const Core::Molecule& molecule,
...
-                   interface.radiusScalar;
-
-    TextLabel3D* atomLabel = createLabel(text, pos, radius, color);
-    geometry->addDrawable(atomLabel);
+    std::string text = "";
+    if (interface.atomOptions & LayerLabel::LabelOptions::Custom) {
+      text += (text == "" ? "" : " / ") + atom.label();
+    }
+    if (interface.atomOptions & LayerLabel::LabelOptions::Index) {
+      text += (text == "" ? "" : " / ") + std::to_string(atom.index());
```
Comment (ghutchis) [inline avogadro/qtplugins/label/label.cpp:312]: `atom.index() + 1` // most users expect Atom 1, Atom 2, etc. rather than starting at zero...

Comment (serk12) [reply avogadro/qtplugins/label/label.cpp:312]: Yes, but the "Select by Atom Index" is indexed at 0. I can change both elements to indexed at 1 if you prefer.

**Comment by serk12** [comment]
Though adding all the combinations looks noisy, if a user needs any of them it will be complex (and probably ugly). I can try a custom sort so it's more appealing or just remove them.

File: avogadro/qtplugins/label/label.cpp (hunk 7 above)
Comment (ghutchis) [reply avogadro/qtplugins/label/label.cpp:316]: Yes, I think most users expect atom numbers to start at 1.

**Comment by ghutchis** [comment]
I think it's a great start. I might tweak the icon a bit (e.g., a letter in front of an atom).

There's a lot of interest in scripts adding labels to atoms, so this is great. Need to add the atom labels to the property table too.

**Pull request merged**